src/components/intent.ts

```diff
@@ -0,0 +1,850 @@
+/*
+Copyright 2020 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import MatrixUser from "../models/users/matrix";
+import JsSdk from "matrix-js-sdk";
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const { MatrixEvent, RoomMember } = JsSdk as any;
+import ClientRequestCache from "./client-request-cache";
+import { defer } from "../utils/promiseutil";
+
+type MatrixClient = {
```

Comment:
My concern with this is what we should probably be using matrix-js-sdk's types. Although I believe they do not exist for the time being, I am worried by strictly typing us to just these functions, anything extra that the bridges may be using (for example, uploadContent) won't be accessible without type hacks.

I am strongly against us doing types for matrix-js-sdk, as they often have breaking releases.

Reply:
`matrix-js-sdk` doesn't currently provide types. These are mainly used internally, but also dictate the return type of any method returning the. 
A lot of methods of intent are just relaying promises from the client.
An alternative would be to import `matrix-js-sdk` (implicitly as `any`) and specifying the returns types explicitly.
Either way, as long as matrix-js-sdk isn't providing types, this file will need to specify some for it.

Yes, the current implementation gives false certainty over functions, parameters and return types that may or may not exist.

Reply:
I think it would be better to just keep it as `any` for now, and wait for types to exist. We can sugar coat it somewhat by having explicit return types on the intent functions.